### PR TITLE
codes

### PR DESCRIPTION
**Pull request by acrantel**

### CODE REVIEW

**Review by 0aax — APPROVED** [review]

**Comment by proan** [comment]
In general, it is good practice to open pull requests sooner, which means breaking up features into smaller units. This makes reviewing easier, and it's less likely to have merge conflicts.

File: src/main/deploy/auton/bezier.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# drive in a spline
+pos 0.0 0.0
+bez 80.0 20.0 -40.0 50.0 25.0 80.0
```
Comment (proan) [inline src/main/deploy/auton/bezier.txt:3]: A comment explaining units or referencing the values would be helpful

File: src/main/deploy/auton/bezier2.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# drive in a spline
+pos 0.0 0.0
+bez 160.0 100.0 -80.0 60.0 40.0 200.0
```
Comment (proan) [inline src/main/deploy/auton/bezier2.txt:3]: Is it possible to put all these paths into a single file? Why are there 3 versions but no comments differentiating each different one?

File: src/main/deploy/auton/paly.txt [hunk 3]
```diff
@@ -0,0 +1,3 @@
+## full field test
```
Comment (proan) [inline src/main/deploy/auton/paly.txt:1]: Name this file better. I think I know what you mean, but what's the problem with a few more characters to make it really clear?

File: src/main/deploy/beta2020.txt [hunk 4]
```diff
@@ -0,0 +1,76 @@
+# Config file for 2020 preseason
+# Domain: roborio-192-frc.local
+
+swerve_width=22.25
```
Comment (proan) [inline src/main/deploy/beta2020.txt:4]: units?

File: src/main/deploy/beta2020.txt [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+# intake motor CAN id
+intake_talon_id=
+intake_sol_id=
+
+# jetson addresses and ports to connect to
+jetson_address=10.1.92.14
+jetson_lidar_port=1030
+jetson_camera_port=1337
+intake_speed=0.9
```
Comment (proan) [inline src/main/deploy/beta2020.txt:55]: Is this related to the Jetson or to the intake motor on L 48–49?

File: src/main/deploy/beta2020.txt [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+#Shooter Values
+shooter_high_angle=50
+shooter_high_angle=25
```
Comment (proan) [inline src/main/deploy/beta2020.txt:59]: Are these supposed to be assigning to the same variable?

File: src/main/deploy/omega2020.txt [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+# intake motor CAN id
+intake_talon_id=6
+intake_sol_id=4
+
+# this is the inital intake speed
+intake_speed=0
+
+#Shooter Values
+shooter_high_angle=50
+shooter_high_angle=25
```
Comment (proan) [inline src/main/deploy/omega2020.txt:58]: Same as in other file. Are these supposed to write to the same variable?

File: src/main/deploy/omega_rpms.txt [hunk 8]
```diff
@@ -0,0 +1,22 @@
+# format: distance (in integer inches), rpm (integer)
```
Comment (proan) [inline src/main/deploy/omega_rpms.txt:1]: thanks for including units!

File: src/main/java/frc/swerve/Wheel.java [hunk 9]
```diff
@@ -105,16 +102,30 @@ public double getCurrentPosition() {
...
+	/** get the drive motor speed in rotations/second */
+	public double getRawDriveSpeed() {
+		// (rotations/minute) * (1 min/60 sec)
+		return driveEncoder.getVelocity() / 60;
+	}
+
+	/** get the rotate motor speed in rotations/sec */
+	public double getRawRotateSpeed() {
+		// (ticks/100ms) / (ticks/rotation) * (10 (100ms)/1s)
+		return (rotateMotor.getSelectedSensorVelocity() / TICKS_PER_ROTATION) * 10;
```
Comment (proan) [inline src/main/java/frc/swerve/Wheel.java:124]: The `10` here comes from the update rate? Can this rate change? I'd recommend a variable or named constant here.

File: src/main/java/frc/positiontracking/PositionTracking.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+        Vector estimate = null;
+        if (estimate != null) {
+            Mat Z = new Mat(STATES, 1, TYPE);
+            Z.put(0, 0, estimate.x, estimate.y);
+            kf.correct(Z);
+        }
+        double tempX = getX();
+        double tempY = getY();
+        double FIELD_HEIGHT = 629.25;
+        double FIELD_WIDTH = 323.25;
```
Comment (proan) [inline src/main/java/frc/positiontracking/PositionTracking.java:101]: These values should be in the configuration file.

File: src/main/java/frc/pathfinding/fieldmap/geometry/VisionTarget.java [hunk 11]
```diff
@@ -0,0 +1,21 @@
+package frc.pathfinding.fieldmap.geometry;
+
+public class VisionTarget {
+
+    private static final double HIGH_HEIGHT = 98.188;
+    private static final double LOW_HEIGHT = 16.5;
```
Comment (proan) [inline src/main/java/frc/pathfinding/fieldmap/geometry/VisionTarget.java:6]: To promote code reuse (i.e. for next year's team), you should set these values in the config file. It's fine to use them below like you have.

File: src/main/java/frc/pathfinding/fieldmap/geometry/Vector.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
+package frc.pathfinding.fieldmap.geometry;
+
+public class Vector {
```
Comment (proan) [inline src/main/java/frc/pathfinding/fieldmap/geometry/Vector.java:3]: This looks pretty good, but consider using a full-blown linear algebra library (such as http://la4j.org) if you need to do any more complicated math.

File: src/main/java/frc/pathfinding/fieldmap/FieldMap.java [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+						new Vector(0, FIELD_SHORT));
+		ROBOT_RADIUS = Math.max(BIGData.getDouble("robot_width"), BIGData.getDouble("robot_height")) / 2;
+	}
+
+	private static void buildMap() {
+		obstacles = new Polygon[7];
+		safeZones = new Polygon[3];
+		visionTargets = new VisionTarget[1];
+		FIELD_SHORT = 323.25;
+		FIELD_LONG = 629.25;
```
Comment (proan) [inline src/main/java/frc/pathfinding/fieldmap/FieldMap.java:34]: This is defined in another file (I just commented on it). Please move all these definitions to a single location.

File: src/main/java/frc/pathfinding/Node.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+        this.pos = pos;
+        neighbors = new HashSet<>();
+    }
+
+    public void update(Node node) {
+        double newG = node.g + distance(node);
+        if (newG < g) {
+            parent = node;
+            g = newG;
+            f = g + h;
```
Comment (proan) [inline src/main/java/frc/pathfinding/Node.java:26]: Better variable names, please.

File: src/main/java/frc/pathfinding/Pathfinding.java [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
+        }
+        if (!isEmergency) {
+            velocity = velocity.multiply(1 / velocity.magnitude());
+        } else {
+            velocity = emergency.multiply(1 / emergency.magnitude());
+        }
+        return velocity;
+    }
+
+    private void initNodes() {
```
Comment (proan) [inline src/main/java/frc/pathfinding/Pathfinding.java:112]: You should separate the algorithm part (above) from the field specifics (below). It makes for easier reuse and easier testing.

File: src/main/java/frc/mechs/StorageMech.java [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+
+        // System.out.println("Top sensor " + topMedVal);
+        // System.out.println("Bot sensor " + botMedVal);
+        // System.out.println("Mid sensor " + midMedVal);
+        // System.out.println("In sensor: " + intakeMedVal);
+
+        // System.out.println("Top sensor " + top.getValue());
+        // System.out.println("Bot sensor " + bottom.getValue());
+        // System.out.println("Mid sensor " + middle.getValue());
+        // System.out.println("Intake sensor " + intake.getValue());
```
Comment (proan) [inline src/main/java/frc/mechs/StorageMech.java:200]: Remove these if they're not necessary (you have git to store your history, after all)

File: src/main/java/frc/control/PathfindingControl.java [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+        }
+    }
+
+    private static void runPFP() {
+        velocity = path.searchPFP(currentPos).multiply(SPEED);
+    }
+
+    private static void runAStar() {
+        dtoNext = currentPos.distanceTo(nextPos);
+        velocity = nextPos.subtract(currentPos).multiply(1 / d).multiply(SPEED);
```
Comment (proan) [inline src/main/java/frc/control/PathfindingControl.java:73]: just to be on the safe side, if you want floating point division, replace `1` with `1.0`. Ask me how I know.

File: src/main/deploy/beta2020.txt (hunk 6 above)
Comment (acrantel) [reply src/main/deploy/beta2020.txt:59]: They should not be, thanks for catching that!

**Pull request merged**